lib/src/license_detection/primary_filter.dart

```diff
@@ -0,0 +1,37 @@
+import 'package:pana/src/license_detection/license.dart';
+
+/// Filters the corpus licenses and
+/// returns a list of [License] which might
+/// be a possible match.
+List<License> filter(Map<String, int> occurences, List<License> knownLicenses) {
+  var possibleLicenses = <License>[];
+
+  knownLicenses.forEach((license) {
+    if (tokenSimilarity(occurences, license.occurences) >= 0.5) {
+      possibleLicenses.add(license);
+    }
+  });
+
+  return List.unmodifiable(possibleLicenses);
+}
+
+/// Returns a measure for token similarity, indicating if [License] is plausibly present as a
+/// substring in [input], as determined by frequency tables over tokens given in [input] and
+/// [knownLicense]
+// Checks if the unknown text contains enough occurences
+// of tokens in a known license to qualify it as a possible match.
+// It there are less number of a particular token in input as compared to known license
+// there is a low probablity that it might match and hence we do not count it, on the other
+// hand if there are more or equal number we count that token
+// and finally return the result (number of qualified unique tokens)/(total number of unique tokens in known license).
+double tokenSimilarity(Map<String, int> input, Map<String, int> knownLicense) {
+  var matches = 0;
+
+  input.keys.forEach((key) {
```

Comment:
Consider using `Map.forEach` see: https://api.dart.dev/stable/2.13.4/dart-core/Map/forEach.html

And maybe using `knownLicense[key] ?? 0` instead of doing `knownLicense.containsKey`...

-------

If we were to have a little fun and be fancy we could something like:

```dart
final matches = input.entries.where((e) => e.value >= knownLicense[e.key] ?? 0).length;
```
Sure, I'm making an iterable of `MapEntry<String,int>`, but I'm only using to call `.length` which counts how many entries it has.

Reply:
Won't this logic also count those tokens which are not present in `knownLicense`?

Reply:
hmm, I think you're right...

hmm... aren't we counting the wrong thing here in the first place too: `/ input.keys.length;` seems wrong :see_no_evil: 

-------

Perhaps it's more like:
```dart

var matches = 0;
knownLicense.forEach((key, value) {
  if (input[key] ?? 0 >= value) {
    matches++;
  }
});

return matches / knownLicense.length;
```

Isn't this what we're supposed to do? Correct me if I'm wrong :smile: 

This could also be written as:
```dart
double tokenSimilarity(Map<String, int> input, Map<String, int> knownLicense) =>
  knownLicense.entries.where((e) => input[e.key] ?? 0 >= e.value).length / knownLicense.length;
```
Or at least I think it I could, right?